apps/omg/lib/omg/fees.ex

```diff
@@ -63,108 +65,65 @@ defmodule OMG.Fees do
+  ...>   }
+  ...> }
+  ...> )
+  %{
+    "eth" => %{
+      amount: 1,
+      subunit_to_unit: 1000000000000000000,
+      pegged_amount: 4,
+      pegged_currency: "USD",
+      pegged_subunit_to_unit: 100,
+      updated_at: DateTime.from_iso8601("2019-01-01T10:10:00+00:00")
+    },
+    "omg" => %{
+      amount: 3,
+      subunit_to_unit: 1000000000000000000,
+      pegged_amount: 4,
+      pegged_currency: "USD",
+      pegged_subunit_to_unit: 100,
+      updated_at: DateTime.from_iso8601("2019-01-01T10:10:00+00:00")
+    }
+  }
 
   """
-  @spec for_transaction(Transaction.Recovered.t(), optional_fee_t()) :: optional_fee_t()
+  @spec for_transaction(Transaction.Recovered.t(), full_fee_t()) :: optional_fee_t()
   def for_transaction(transaction, fee_map) do
     case MergeTransactionValidator.is_merge_transaction?(transaction) do
       true -> :no_fees_required
-      false -> fee_map
+      false -> get_fee_for_type(transaction, fee_map)
     end
   end
 
-  @doc ~S"""
-  Returns a filtered map of fees given a list of desired currencies.
-  Fees will not be filtered if an empty list of currencies is given.
-
-  ## Examples
-
-      iex> OMG.Fees.filter_fees(
-      ...> %{
-      ...>  "eth" => %{
-      ...>    amount: 1,
-      ...>    subunit_to_unit: 1000000000000000000,
-      ...>    pegged_amount: 4,
-      ...>    pegged_currency: "USD",
-      ...>    pegged_subunit_to_unit: 100,
-      ...>    updated_at: DateTime.from_iso8601("2019-01-01T10:10:00+00:00")
-      ...>  },
-      ...>  "omg" => %{
-      ...>    amount: 3,
-      ...>    subunit_to_unit: 1000000000000000000,
-      ...>    pegged_amount: 4,
-      ...>    pegged_currency: "USD",
-      ...>    pegged_subunit_to_unit: 100,
-      ...>    updated_at: DateTime.from_iso8601("2019-01-01T10:10:00+00:00")
-      ...>  }
-      ...> },
-      ...> ["eth"]
-      ...> )
-      {:ok,
-        %{
-          "eth" =>
-          %{
-            amount: 1,
-            subunit_to_unit: 1000000000000000000,
-            pegged_amount: 4,
-            pegged_currency: "USD",
-            pegged_subunit_to_unit: 100,
-            updated_at: DateTime.from_iso8601("2019-01-01T10:10:00+00:00")
-          }
-        }
-      }
-
-  """
-  @spec filter_fees(fee_t(), list(String.t()) | nil) :: {:ok, fee_t()} | {:error, :currency_fee_not_supported}
-  # empty list = no filter
-  def filter_fees(fees, []), do: {:ok, fees}
-  def filter_fees(fees, nil), do: {:ok, fees}
-
-  def filter_fees(fees, desired_currencies) do
-    Enum.reduce_while(desired_currencies, {:ok, %{}}, fn currency, {:ok, filtered_fees} ->
-      case Map.fetch(fees, currency) do
-        :error -> {:halt, {:error, :currency_fee_not_supported}}
-        {:ok, fee} -> {:cont, {:ok, Map.put(filtered_fees, currency, fee)}}
-      end
-    end)
+  defp get_fee_for_type(%Transaction.Recovered{signed_tx: %Transaction.Signed{raw_tx: raw_tx}}, fee_map) do
+    case WireFormatTypes.module_tx_types()[raw_tx.__struct__] do
```

Comment:
A bit worried that this reliance on `__struct__` will be hard to spot & update if the behavior changes in the future.

Maybe convert this to e.g. `WireFormatTypes.tx_type_from_transaction()`?